### ISSUE TITLE
Principals of type Application cannot validly be used in role assignments.

### ISSUE THREAD

**Issue by rnkhouse** [Azure/terraform-azurerm-appgw-ingress-k8s-cluster#1]
Getting below error on `terraform apply`:

    2 error(s) occurred:

    * azurerm_role_assignment.ra1: 1 error(s) occurred:

    * azurerm_role_assignment.ra1: authorization.RoleAssignmentsClient#Create: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="PrincipalTypeNotSupported" Message="Principals of type Application cannot validly be used in role assignments."
    * azurerm_kubernetes_cluster.test: 1 error(s) occurred:

    * azurerm_kubernetes_cluster.test: Error creating/updating Managed Kubernetes Cluster "aks-cluster1" (Resource Group "aksrg1"): containerservice.ManagedClustersClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="ServicePrincipalNotFound" Message="Service principal clientID: xxxxx-xxxxx-xxxxx-xxxxx not found in Active Directory tenant xxxxx-xxxxx-xxxxx-xxxxx, Please see https://aka.ms/acs-sp-help for more details."

**Comment by alanold** [comment]
I'm getting the same error (Principals of type Application cannot validly be used in role assignments) when trying to add deploy role assignments via ARM template

I've tried parsing the appId and objectId of the service principle 
Any update on this ?

**Comment by dabasejumper** [comment]
Getting the same errors as well.

**Comment by solomson** [comment]
Same issue in ARM template

**Comment by uolter** [comment]
I had the same issue. Paying more attention to the name of the variable in my terraform.tfvars definition I realized I was using the wrong object_id. Indeed it is required the one related to my subscription instead of the application object id.
I hope this is helpful.

**Comment by christian-vorhemus** [comment]
And as an additional input for everyone else stumbling into this: When you create a new "app registration" in the Azure portal, actually two objects are created: An application object and a service principal object. The object ID which appears in the Azure portal is the application object ID, not the service principal object ID. So what you can do is:

Create a SP:
`az ad sp create-for-rbac -n "sample-sp"`

List details for the newly created SP:
`az ad sp list --filter "displayName eq 'sample-sp'"`

![objectid](https://user-images.githubusercontent.com/3229424/72366525-b6833f00-36fa-11ea-959e-076d66604246.png)

Use the object ID from the CLI as the object ID you enter in Terraform.

**Comment by RobertoBorges** [comment]
> And as an additional input for everyone else stumbling into this: When you create a new "app registration" in the Azure portal, actually two objects are created: An application object and a service principal object. The object ID which appears in the Azure portal is the application object ID, not the service principal object ID. So what you can do is:
> 
> Create a SP:
> `az ad sp create-for-rbac -n "sample-sp"`
> 
> List details for the newly created SP:
> `az ad sp list --filter "displayName eq 'sample-sp'"`
> 
> ![objectid](https://user-images.githubusercontent.com/3229424/72366525-b6833f00-36fa-11ea-959e-076d66604246.png)
> 
> Use the object ID from the CLI as the object ID you enter in Terraform.

So why not to show in the portal the Service Principal Object ID ?
Every day many folks have the same problem not just with AKS but with many other services in Azure!

**Comment by severity1** [comment]
You can also find the object id through the Enterprise Application.

**Comment by thenakulchawla** [comment]
> And as an additional input for everyone else stumbling into this: When you create a new "app registration" in the Azure portal, actually two objects are created: An application object and a service principal object. The object ID which appears in the Azure portal is the application object ID, not the service principal object ID. So what you can do is:
> 
> Create a SP: `az ad sp create-for-rbac -n "sample-sp"`
> 
> List details for the newly created SP: `az ad sp list --filter "displayName eq 'sample-sp'"`
> 
> ![objectid](https://user-images.githubusercontent.com/3229424/72366525-b6833f00-36fa-11ea-959e-076d66604246.png)
> 
> Use the object ID from the CLI as the object ID you enter in Terraform.

I tried this, but there was no key `ObjectId` in my json output.

**Comment by ArnWEB** [comment]
same issue

**Comment by thenakulchawla** [comment]
> same issue

I found the answer. The json response has an `id` key, which is actually the `objectId`.

**Comment by alexivanov-ai** [comment]
I have to agree this is missleading. Application_id i understand, but ojbect_id i don't. And things changed, as this requires the object_id (`id`) or the service principal runnning behind the application (App registration)
https://learn.microsoft.com/en-us/cli/azure/role/assignment?view=azure-cli-latest#az-role-assignment-create-optional-parameters: 
`--assignee-object-id
Use this parameter instead of '--assignee' to bypass Graph API invocation in case of insufficient privileges. This parameter only works with object ids for users, groups, service principals, and managed identities. For managed identities use the principal id. For service principals, use the object id and not the app id.`

**Comment by PLsergent** [comment]
I don't know if it's going to be useful but after creating an app registration you can access the service principal object Id from the enterprise application and this solved it for me. The object id of the enterprise application is different from the one in the app registration.
![image](https://github.com/Azure/terraform-azurerm-appgw-ingress-k8s-cluster/assets/27278774/58be8f25-7961-44c5-967a-b3acce167af5)
![image](https://github.com/Azure/terraform-azurerm-appgw-ingress-k8s-cluster/assets/27278774/4ebd501a-8fa1-4556-9812-1e735e1adf32)
